### PR TITLE
Added lint descriptions

### PR DESCRIPTION
**Pull request by TheMC47**

### CODE REVIEW

**Comment by TheMC47** [comment]
As of now, running `isabelle lint_descriptions` produces the following table:

<table>
<tr><th>Name</th><th>Severity</th><th>Description</th></tr>
<tr><td>apply_isar_switch</td><td>warn</td><td>Switching form an apply script to a structured Isar proof results in an overall proof that is hard to read without relying on Isabelle. The Isar proof is also sensitive to the output of the apply script, and might therefore break easily. This lint detects these switches.<br /><br />

References: http://proofcraft.org/blog/isabelle-style.html</td></tr>
<tr><td>auto_structural_composition</td><td>info</td><td>Using <code>apply (auto;…)</code> ) results in a behavior that is hard to predict, so it is discouraged. This lint informs users about this problem.<br /><br />

References: http://proofcraft.org/blog/isabelle-style-part2.html</td></tr>
<tr><td>axiomatization_with_where</td><td>error</td><td>Unless when creating a new logic or extending an existing one with new axioms,the <code>axiomatization</code> command, when used, should not include a <code>where</code> clause.<br /><br />
The problem with the <code>where</code> clause is that it can introduce inconsistenciesinto the logic, for example:

```isabelle
axiomatization
  P :: "'a ⇒ bool"
where
  all_true: "∀x. P x" and
  all_false: "∀x. ¬P x"

```
This lint finds such problematic  <code>axiomatization</code> invocations.<br /><br />

References: http://proofcraft.org/blog/isabelle-style.html</td></tr>
<tr><td>bad_style_command</td><td>warn</td><td>This lint detects bad-style commands:
 <code>back, apply_end</code> </td></tr>
<tr><td>complex_isar_initial_method</td><td>warn</td><td>Initial  <code>proof</code> methods should be kept simple, in orderto keep the goals of the proof clear. For example, simplifier calls shouldbe avoided, and not many methods should be combined. This lint finds complexmethods in proof commands.<br /><br />

References: http://proofcraft.org/blog/isabelle-style.html</td></tr>
<tr><td>complex_method</td><td>warn</td><td>This lint warns users from using complex method. A method is considered complex,when one of the following conditions hold:

- has more than one modifier (?, +, or []), for example  <code>auto?[4]</code>
- has modifiers that are not at the outmost level, for example  <code>auto[3] | blast</code>
- has three or more combinators (|, ;, ,), for example  <code>auto ; rule , (force | blast)</code>
</td></tr>
<tr><td>counter_example_finder</td><td>error</td><td>This lint detects counter-example finders:
 <code>nitpick, nunchaku, quickcheck</code> </td></tr>
<tr><td>diagnostic_command</td><td>info</td><td>This lint finds diagnostic commands:
 <code>ML_val, class_deps, code_deps, code_thms, find_consts, find_theorems, find_unused_assms, full_prf, help, locale_deps, prf, print_ML_antiquotations, print_abbrevs, print_antiquotations, print_attributes, print_bnfs, print_bundles, print_case_translations, print_cases, print_claset, print_classes, print_codeproc, print_codesetup, print_coercions, print_commands, print_context, print_definitions, print_defn_rules, print_facts, print_induct_rules, print_inductives, print_interps, print_locale, print_locales, print_methods, print_options, print_orders, print_quot_maps, print_quotconsts, print_quotients, print_quotientsQ3, print_quotmapsQ3, print_record, print_rules, print_simpset, print_state, print_statement, print_syntax, print_term_bindings, print_theorems, print_theory, print_trans_rules, smt_status, thm_deps, thm_oracles, thy_deps, unused_thms, value, values, welcome, term, prop, thm, typ</code> </td></tr>
<tr><td>force_failure</td><td>info</td><td>Since some methods do not guarantee to solve all their goals, it might be helpful to consider forcing their failure (e.g. using <code>apply (simp; fail)</code> instead of just <code>apply simp</code> in order to make debugging proofs easier. This lint tries to find these methods.<br /><br />

References: http://proofcraft.org/blog/isabelle-style-part2.html</td></tr>
<tr><td>global_attribute_changes</td><td>info</td><td>Changing lemma attributes (e.g. <code>simp</code> ) to be used in a local proof isdiscouraged, as it is error-prone and might result in hard-to debug difficulties problems if the lemma should not be globally used as a simplifier rule or when there are some conflicts with other theories.
<br /><br />
Concretely, the lints warns the users of using this pattern:

```isabelle
declare word_neq_0_conv [simp]

lemma …
lemma …

declare word_neq_0_conv [simp del]

```
Instead, users should use the context, notes or bundle commands.<br /><br />

References: http://proofcraft.org/blog/isabelle-style-part2.html</td></tr>
<tr><td>global_attribute_on_unnamed_lemma</td><td>error</td><td>Setting a global attribute (like <code>simp</code> or <code>elim</code> ) on an unnamed lemma should be avoided, since it can make debugging proofsremoving the effect of that attribute harder. This lint finds such lemmas.<br /><br />

References: http://proofcraft.org/blog/isabelle-style.html</td></tr>
<tr><td>implicit_rule</td><td>warn</td><td>Using <code>apply rule</code> results in Isabelle finding the suitable rule for the given context. However, if the process for finding the rule changes in the future, the proof might break. Instead, users should explicitly state which rule is needed. This lint finds such invocations.<br /><br />

References: http://proofcraft.org/blog/isabelle-style.html</td></tr>
<tr><td>lemma_transforming_attribute</td><td>warn</td><td>This lint warns of using transforming attributes ( <code>simplified, rule_format, and unfolded</code> ) on lemmas. Instead, the user should write the transformed form.<br /><br />

References: http://proofcraft.org/blog/isabelle-style-part2.html</td></tr>
<tr><td>low_level_apply_chain</td><td>info</td><td>Using long apply-scripts with low methods can quickly make proofs unreadable and unnecessarily long. This lints flags such scripts that are longer than 5 commands.</td></tr>
<tr><td>proof_finder</td><td>error</td><td>This lint detects proof-finder commands:
 <code>sledgehammer, solve_direct, try, try0</code> </td></tr>
<tr><td>short_name</td><td>info</td><td>Finds functions or definitions with short names (one character).</td></tr>
<tr><td>unfinished_proof</td><td>error</td><td>This lint detects unfinished proofs, characterized by the following commands:
 <code>sorry, oops, \<proof></code> </td></tr>
<tr><td>unrestricted_auto</td><td>error</td><td>Using auto in the middle of a proof on all goals (i.e. unrestricted) might produce an unpredictable proof state. It should rather be used as a terminal proof method, or be restricted to a set of goals that it fully solves. The lint finds unrestricted auto invocations that are not terminal.
<br /><br />

References: http://proofcraft.org/blog/isabelle-style.html</td></tr>
<tr><td>use_by</td><td>info</td><td>The <code>by</code> command allows to express method applications using <code>apply</code> more concisely. For example, instead of

```isabelle
lemma …
  apply (induction xs)
  apply auto
  done
```

 <code>by</code> can be used:

```isabelle
lemma …
  by (induction xs) auto
```

The lint identifies valid replacements and suggests them.</td></tr>
<tr><td>use_isar</td><td>info</td><td>This lint triggers on every use of the  <code>apply</code> command and suggests to use an Isar proof instead.</td></tr>
</table>

**Comment by TheMC47** [comment]
(I need to squash the commits before merging)

**Comment by TheMC47** [comment]
closes #2

File: linter_base/src/lints.scala [hunk 1]
```diff
@@ -345,7 +472,8 @@ object Unfinished_Proof
     "Consider finishing the proof.",
     "unfinished_proof",
     List("sorry", "oops", "\\<proof>"),
```
Comment (Dacit) [inline linter_base/src/lints.scala:474]: Does not render properly

File: linter_base/src/lints.scala [hunk 2]
```diff
@@ -308,6 +401,31 @@ object Axiomatization_With_Where extends Single_Command_Lint
...
+      .add("The problem with the ").keyword("where").add(" clause is that it can introduce inconsistencies")
+      .addln("into the logic, for example:")
+      .code_block(
+        "axiomatization",
+        """  P :: "'a ⇒ bool"""",
+        "where",
+        """  all_true: "∀x. P x" and """,
+        """all_false: "∀x. ¬P x""""
+      )
+      .add("This lint finds such problematic ").keyword("axiomatization").add(" invocations.")
```
Comment (Dacit) [inline linter_base/src/lints.scala:424]: Remove redundant line

File: linter_base/src/print_lints.scala [hunk 3]
```diff
@@ -0,0 +1,67 @@
+package isabelle.linter
+
+
+import isabelle._
+import linter.Linter._
+
+object Lint_Descriptions
+{
+
+  private def surround(str: String, kind: String): String
```
Comment (Dacit) [inline linter_base/src/print_lints.scala:10]: `str` needs to be escaped (see render problem of unfinished_proof link)

File: linter_base/src/lints.scala [hunk 4]
```diff
@@ -11,6 +11,19 @@ object Apply_Isar_Switch extends Proper_Commands_Lint
...
+  val short_description: Lint_Description =
+    Lint_Description.empty
+      .add("Switching form an apply script to a structured Isar proof is error-prone and hard to read.")
+
+  val long_description: Lint_Description =
+    Lint_Description.empty
+      .add("Switching form an apply script to a structured Isar proof results in an overall ")
+      .add("proof that is hard to read without relying on Isabelle. The Isar proof is also ")
+      .add("sensitive to the output of the apply script, and might therefore break easily. ")
+      .add("This lint detects these switches.")
```
Comment (Dacit) [inline linter_base/src/lints.scala:23]: Remove redundant line

File: linter_base/src/lints.scala [hunk 5]
```diff
@@ -573,6 +773,16 @@ object Auto_Structural_Composition extends AST_Lint
...
   val severity: Severity.Level = Severity.Info
 
+  val short_description: Lint_Description =
+    Lint_Description.empty
+      .add("Using ").keyword("apply (auto;…)").add(" is discouraged.")
+
+  val long_description: Lint_Description =
+    Lint_Description.empty
+      .add("Using ").keyword("apply (auto;…)").add(" results in a behavior that is hard to predict, ")
+      .add("so it is discouraged. This lint informs users about this problem.")
```
Comment (Dacit) [inline linter_base/src/lints.scala:783]: Remove redundant "This lint informs users about this problem."

File: linter_base/src/lints.scala [hunk 6]
```diff
@@ -534,6 +708,19 @@ object Complex_Isar_Initial_Method extends AST_Lint
...
   val severity: Severity.Level = Severity.Warn
 
+  val short_description: Lint_Description =
+    Lint_Description.empty
+      .add("Using complex methods in the ").keyword("proof").add(" command makes the proof")
+      .add(" brittle and hard to read.")
+
+  val long_description: Lint_Description =
+    Lint_Description.empty
+      .add("Initial ").keyword("proof").add(" methods should be kept simple, in order")
```
Comment (Dacit) [inline linter_base/src/lints.scala:718]: Missing space

File: linter_base/src/lints.scala [hunk 7]
```diff
@@ -534,6 +708,19 @@ object Complex_Isar_Initial_Method extends AST_Lint
...
+      .add("to keep the goals of the proof clear. For example, simplifier calls should")
```
Comment (Dacit) [inline linter_base/src/lints.scala:719]: Missing space

File: linter_base/src/lints.scala [hunk 8]
```diff
@@ -534,6 +708,19 @@ object Complex_Isar_Initial_Method extends AST_Lint
...
+      .add("be avoided, and not many methods should be combined. This lint finds complex")
```
Comment (Dacit) [inline linter_base/src/lints.scala:720]: Missing space

File: linter_base/src/lints.scala [hunk 9]
```diff
@@ -596,6 +806,23 @@ object Complex_Method extends AST_Lint
   val name: String = "complex_method"
   val severity: Severity.Level = Severity.Warn
 
+  val short_description: Lint_Description =
+    Lint_Description.empty
+      .add("Using complex methods (e.g. too much nesting) should be avoided.")
+
+  val long_description: Lint_Description =
+    Lint_Description.empty
+      .add("This lint warns users from using complex method. A method is considered complex,")
```
Comment (Dacit) [inline linter_base/src/lints.scala:815]: More concise: "Warns users from using overly complex methods, i.e. if either of the following holds: "

File: linter_base/src/lints.scala [hunk 10]
```diff
@@ -596,6 +806,23 @@ object Complex_Method extends AST_Lint
...
+      .addln("when one of the following conditions hold:")
+      .breakline
+      .add("- has more than one modifier (?, +, or []), for example ").keyword("auto?[4]")
```
Comment (Dacit) [inline linter_base/src/lints.scala:818]: Code-set `(?, +, or []`?

File: linter_base/src/lints.scala [hunk 11]
```diff
@@ -596,6 +806,23 @@ object Complex_Method extends AST_Lint
...
+      .breakline
+      .add("- has modifiers that are not at the outmost level, for example ").keyword("auto[3] | blast")
+      .breakline
+      .add("- has three or more combinators (|, ;, ,), for example ").keyword("auto ; rule , (force | blast)")
```
Comment (Dacit) [inline linter_base/src/lints.scala:822]: Code-set `(|, ;, ,)`?

File: linter_base/src/lints.scala [hunk 12]
```diff
@@ -560,6 +747,19 @@ object Force_Failure extends AST_Lint
...
+    Lint_Description.empty
+      .add("Forcing failure, for example ").keyword("apply (simp; fail)")
+      .add(", might be helpful.")
+
+  val long_description: Lint_Description =
+    Lint_Description.empty
+      .add("Since some methods do not guarantee to solve all their goals, it might be ")
+      .add("helpful to consider forcing their failure (e.g. using ")
+      .keyword("apply (simp; fail)").add(" instead of just ").keyword("apply simp")
+      .add(") in order to make debugging proofs easier. This lint tries to find these methods.")
```
Comment (Dacit) [inline linter_base/src/lints.scala:760]: Remove redundant "This lint tries to find these methods"

File: linter_base/src/lints.scala [hunk 13]
```diff
@@ -497,6 +646,18 @@ object Lemma_Transforming_Attribute extends Parser_Lint
...
 
+  val short_description: Lint_Description =
+    Lint_Description.empty
+      .add("Setting lemma-transforming attributes on lemmas should be avoided,")
+      .add(" instead the transformed form should be used.")
+
+  val long_description: Lint_Description =
+    Lint_Description.empty
+      .add("This lint warns of using transforming attributes (")
+      .keyword("simplified, rule_format, and unfolded")
```
Comment (Dacit) [inline linter_base/src/lints.scala:657]: multiple keywords

File: linter_base/src/lint_description.scala [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+    description.rev_els.reverse.map(render).mkString
+}
+
+object Markdown_Renderer extends Lint_Description_Renderer
+{
+  def wrap(left: String, right: String)(content: String): String =
+    s"$left$content$right"
+
+  def wrapTags(tag: String): String => String =
+    wrap(s"<$tag>", s"</$tag>")
```
Comment (Dacit) [inline linter_base/src/lint_description.scala:62]: Code Style comment: Isabelle/Scala does not use string interpolations

**Comment by Dacit** [comment]
Two things I couldn't pin down exactly:
- escaping is off at some point (see `unfinished_proof` description)
- the commands of a proper_command lint should be individually printed as code

**Review by Dacit — COMMENTED** [review]
See discussion comment

File: linter_base/src/lints.scala [hunk 15]
```diff
@@ -220,6 +278,32 @@ object Global_Attribute_Changes extends Proper_Commands_Lint with TokenParsers
...
   val severity: Severity.Level = Severity.Info
 
+  val short_description: Lint_Description =
+    Lint_Description.empty
+      .add("Global lemma attributes should not be changed temporarily, use ")
+      .keyword("declare").add(" instead.")
+
+  val long_description: Lint_Description =
+    Lint_Description.empty
+      .add("Changing lemma attributes (e.g. ").keyword("simp").add(") to be used in a local proof is")
```
[the remark this answers is not in the document]
Comment (TheMC47) [reply linter_base/src/lints.scala:288]: should I keep the example?

File: linter_base/src/lints.scala (hunk 1 above)
Comment (TheMC47) [reply linter_base/src/lints.scala:474]: I think it does, it's supposed to render as `\<proof>`

File: linter_base/src/lints.scala [hunk 16]
```diff
@@ -34,6 +47,30 @@ object Use_By extends Proper_Commands_Lint with TokenParsers
...
+    Lint_Description.empty
+      .add("Using the ").keyword("by")
+      .add(" command is more concise than short apply-scripts.")
+
+  val long_description: Lint_Description =
+    Lint_Description.empty
+      .add("The ").keyword("by")
+      .add(" command allows to express method applications using ").keyword("apply")
+      .addln(" more concisely. For example, instead of")
+      .code_block(
```
[the remark this answers is not in the document]
Comment (TheMC47) [reply linter_base/src/lints.scala:60]: I guess I'll keep it then

**Comment by TheMC47** [comment]
Alright, all taken care of instead of three comments

**Pull request merged**